tests/Middleware/TrustedHostsNetworkResolverTest.php

```diff
@@ -0,0 +1,191 @@
+<?php
+
+namespace Yiisoft\Yii\Web\Tests\Middleware;
+
+use Nyholm\Psr7\Factory\Psr17Factory;
+use Nyholm\Psr7\ServerRequest;
+use PHPUnit\Framework\TestCase;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\MiddlewareInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+use Yiisoft\Yii\Web\Middleware\TrustedHostsNetworkResolver;
+use Yiisoft\Yii\Web\Tests\Middleware\Mock\MockRequestHandler;
+
+class TrustedHostsNetworkResolverTest extends TestCase
+{
+    protected function newRequestWithSchemaAndHeaders(
+        string $scheme = 'http',
+        array $headers = [],
+        array $serverParams = []
+    ): ServerRequestInterface {
+        $request = new ServerRequest('GET', '/', $headers, null, '1.1', $serverParams);
+        $uri = $request->getUri()->withScheme($scheme);
+        return $request->withUri($uri);
+    }
+
+    public function trustedDataProvider(): array
+    {
+        return [
+            'xForwardLevel1' => [
+                ['x-forwarded-for' => ['9.9.9.9', '5.5.5.5', '2.2.2.2']],
+                ['REMOTE_ADDR' => '127.0.0.1'],
+                [['hosts' => ['8.8.8.8', '127.0.0.1']]],
+                '2.2.2.2',
+            ],
+            'xForwardLevel2' => [
+                ['x-forwarded-for' => ['9.9.9.9', '5.5.5.5', '2.2.2.2']],
+                ['REMOTE_ADDR' => '127.0.0.1'],
+                [['hosts' => ['8.8.8.8', '127.0.0.1', '2.2.2.2']]],
+                '5.5.5.5',
+            ],
+            'forwardLevel1' => [
+                ['forward' => ['for=9.9.9.9', 'for=5.5.5.5', 'for=2.2.2.2']],
+                ['REMOTE_ADDR' => '127.0.0.1'],
+                [['hosts' => ['8.8.8.8', '127.0.0.1']]],
+                '2.2.2.2',
```

Comment:
if RFC literally says that it MUST BE appended always, why do we even test for addresses further than first one?

Reply:
https://tools.ietf.org/html/rfc7239#section-8.1

>  The "Forwarded" HTTP header field cannot be relied upon to be
   correct, as it may be modified, whether mistakenly or for malicious
   reasons, by every node on the way to the server, including the client
   making the request.